wxget, wxdocs and wxdemo to scripts (#437)

GadgetSteve opened the pull request:
As discussed in #437 add new wxdemo & wxdoc launchers to scripts, these will ensure that the matching documents or demos for the loaded version of wx are available, downloading and installing them if required and then launch the  demo or the docs.

Both scripts are actually entry points for `wx/tools/wxget_docs_demo.py` which can be invoked directly with a parameter of either `docs` or `demo`.

To implement the fetching of the downloads under windows, which doesn't have wget, without introducing a dependency on `requests` it was necessary to implement a somewhat limited wxget which is usable as a library and as a command line script. This is implemented in `wx/tools/wxget.py`

Fixes #437

RobinD42 commented on wx/tools/wxget_docs_demo.py:
```diff
@@ -0,0 +1,175 @@
+WXVERSION = wx.version().split(' ')[0]
+MAJOR = WXVERSION.split('.')[0]
+if MAJOR != '4':
+    raise ValueError("wx Versions other than 4 not supported!")
```
`wx.VERSION` is a tuple containing the version number components. Using it will save a couple lines here. Also, there is `wx.VERSION_STRING` for the string version of the full version number.

GadgetSteve replied:
Done - much cleaner!

RobinD42 commented on wx/tools/wxget_docs_demo.py:
```diff
@@ -0,0 +1,175 @@
+if MAJOR != '4':
+    raise ValueError("wx Versions other than 4 not supported!")
+
+def endsure_wx_app():
```
"endsure" --> "ensure" typo?

GadgetSteve replied:
Done!

RobinD42 commented on wx/tools/wxget_docs_demo.py:
```diff
@@ -0,0 +1,175 @@
+
+    cached = os.path.join(cache, name)
+    cached = os.path.extsep.join([cached, ext])
+    print('Looking for cashed', cached)
```
"cashed" --> "cached"

RobinD42 commented on wx/tools/wxget_docs_demo.py:
```diff
@@ -0,0 +1,175 @@
+    global APP
+    if APP:
+        print("Closing Launcher App!")  # Debug
+        APP.Destroy()
```
`wx.App.Destroy` can sometimes be unsafe to call, however since you are never calling `MainLoop` then it is sufficient to simply exit normally without cleaning up the app object. If you run into situations where that is not enough then using `wx.Exit()` to force it would be a better way to go.

GadgetSteve replied:
Changed to APP.Exit() - I have had a few cases where the main loop did not seem to be tidied up so thing that this will be safer.

GadgetSteve replied:
Opps - wxExit not APP.Exit!

RobinD42 commented on wx/tools/wxget_docs_demo.py:
```diff
@@ -0,0 +1,175 @@
+        print("Show Docs at:", location)
+        webbrowser.open(location_url)
+    else:
+        result = 'Unable to find wx.Docs'
```
Replace "wx.Docs" with "wxPython Documentation" in the error message.

GadgetSteve replied:
Done

RobinD42 commented on wx/tools/wxget_docs_demo.py:
```diff
@@ -0,0 +1,175 @@
+        #subprocess.check_call(cmds) # Use instead for debug
+        print("Demo starting as PID %s - may take a few seconds!" % pid)
+    else:
+        result = 'Unable to find wx.Demo'
```
Replace "wx.Demo" with "wxPython Demo" in the error message

GadgetSteve replied:
Done

RobinD42 commented on wx/tools/wxget.py:
```diff
@@ -0,0 +1,168 @@
+            dest_dir, filename = os.path.split(dlg.GetPath())
+        else:
+            url = None
+        del dlg
```
`del dlg` is not enough, you need to actually call `dlg.Destroy()`. Or you can use a context manager, which will call `Destroy` for you on the way out, like this:

```python
    with wx.FileDialog(...) as dlg:
        if dlg.ShowModal() == ...
```

GadgetSteve replied:
Dpne

RobinD42 commented:
Thanks!

In addition to the inline comments, please add a note about this to `CHANGES.txt`. 

It would probably also be a good idea to add something about these new scripts to `LONG_DESCRIPTION` in `setup.py`, so people who read the PyPI description will know about them. Similarly, adding something to `docs/sphinx/rest_substitutions/overviews/index.rst` will help people reading the docs online to fetch their own copy for local/off-line reading.

GadgetSteve commented:
Also added a yes/no prompt before going on-line for people who are on monitored or charged connections to have the option of saying NO!

GadgetSteve commented:
Added these items to CHANGES.rst, (since wx/py/CHANGES.txt hasn't been updated since 0.9.7.9 (9/8/09)).

Added to the LONG_DESCRIPTION as suggested.

GadgetSteve commented:
Unable to get Sphinx to pick up these new modules correctly so simply added a note to `docs/sphinx/rest_substitutions/overviews/index.rst` - also added some handling of download errors, e.g. due to firewalls by giving the user the download URL and target location and suggesting using a browser.
